_posts/2022-02-17-passwordless-auth-with-express-and-postgres.md

```diff
@@ -0,0 +1,563 @@
+  }
+});
+```
+
+Now that we're sure the OTP in the database is valid we can compare it to the one provided by the user. Remember to clean up the database by deleting the code in the case that it's used successfully.
+
+```js
+router.post("/token", async (req, res) => {
+  ...
+  
+  // Verify the code matches
+  const match = bcrypt.compareSync(code, otp.code);
+
+  if (!match) throw { message: "Error with code" };
+  
+  // Delete code
+  await Queries.Otp.Delete.ByUserId(user.id);
+});
+```
+
+If we've made it this far the code provided matches the one in the database, and we can log our user in. This is where we use the `makeToken` function we created earlier to generate a JWT with our user's details.
+
+```js
+router.post("/token", async (req, res) => {
+  ...
+  
+  const token = makeToken({
+    ...user,
+  });
+});
+```
+
+We have our secure token! Now we just need to store it in the user's cookies under the name `access_token` and redirect them to the `/account` page.
+
+```js
+router.post("/token", async (req, res) => {
+  ...
+  
+  return res
+    .cookie("access_token", token, {
+      httpOnly: true,
+      secure: true,
+    })
+    .status(200)
+    .redirect("/");
+});
+```
+
+### GET /account
+
+Our user now has a secure JWT verifying their identity. Our next step is to validate that JWT and present the user with a page that shows they are logged in.
+
+First we'll create the middleware function to validate the token so that we can apply it to any endpoint we wish to be protected. The steps here are as follows:
+
+1. Make sure a token exists
+2. Verify that it's not expired
+3. Populate the user data into the request object
+4. Continue to the next step in the application
+
+If any of this fails we want to redirect the user to the login screen.
+
+```js
+const validateToken = (req, res, next) => {
+  const token = req.cookies.access_token;
+
+  try {
+    if (!token) throw true;
+    const data = jwt.verify(token, process.env.JWT_SECRET_KEY);
+    if (!data.expirationDate || Date.parse(data.expirationDate) < Date.now()) {
+      throw true;
+    }
+    req.user = {};
+    Object.keys(data).forEach((key) => {
+      req.user[key] = data[key];
+    });
+    return next();
+  } catch {
+    return res.redirect("/users/login");
+  }
+};
+```
+
+Now that we have the middleware in place we just need to use it to protect an endpoint like so:
+
+```js
+router.get("/account", validateToken, async (req, res) => {
+  res.send(`Hello ${req.user.name}!`);
+});
+```
+
+This endpoint is now protected by a valid JWT, and will only render for a securely logged in user.
+
+We are almost done, just one last piece of the puzzle to go.
+
+### GET /logout
+
+This is the simplest of all the endpoints. All we want to do is clear out the token from the cookie, and redirect them to the login page.
+
+```js
+router.get("/logout", async (req, res) => {
+  return res.clearCookie("access_token").status(200).redirect("/login");
+});
+```
+
+# Summary
```

Comment:
Suggestion; maybe a screenshot of the end result from the user's perspective?

Reply:
I considered that, but the app is so simple it would just be "Hello {email}" so there isn't much to look at unfortunately.